### ISSUE TITLE
[Swarm mode] Plans for auto-configuration based on service labels?

### ISSUE THREAD

**Issue by bvis** [vfarcic/docker-flow-proxy#28]
Hi, 

First of all, thanks for your work, it's very useful and inspiring.

I have a question/suggestion about a useful feature for Swarm mode (Docker 1.12+)

One option that would be very interesting would be about the usage of labels in the service to define some characteristics like:

- I want to be accessible from the 80/443 port
- I want to get traffic from this path and/or this domain.

For example, if I launch a service with something like:
```
docker service create \
  --label "com.proxy.flow.http=80" \
  --label "com.proxy.flow.path=/" \
  --label "com.proxy.flow.domain=go-demo.com" \
  -p 80 \
   --name nginx-public \
  nginx
```
Then, in a similar way consul-template does over consul events, the docker-flow-proxy or a complementary service hearing for docker events could catch those services with these labels and automatically publish the service named *nginx-public* for the go-demo.com domain under the "/" path.

This would allow to launch many services in Docker swarm with just one load balancer over the entire cluster.

An important point is that when a new instance of the proxy is started it should do an initial check of the services that have any of its labels, to initialise the proxy status.

With the current configuration based on the API call (/v1/docker-flow-proxy/reconfigure) you don't have a persistent status, then when I launch more instances of the service I need to connect to each one individually and initialise the configuration.

Another improvement to the service would be to have a "HEALTHCHECK" command in the Dockerfile that would indicate when I launch a new instance of the service, that it has read the initial configuration and is ready for use, therefore it can be included in route mesh table.

**Comment by vfarcic** [comment]
I like the idea of listening to Swarm events and reconfiguring the proxy based on labels. I'll start working on it soon.

> With the current configuration based on the API call (/v1/docker-flow-proxy/reconfigure) you don't have a persistent status, then when I launch more instances of the service I need to connect to each one individually and initialise the configuration.

There is a persistent status (either in Consul or as a file). There is no need to do anything when you launch more instances of an existing service. The proxy needs to be reconfigured only once per each service (not instance nor release).

> Another improvement to the service would be to have a "HEALTHCHECK" command in the Dockerfile that would indicate when I launch a new instance of the service, that it has read the initial configuration and is ready for use, therefore it can be included in route mesh table.

Swarm already uses HEALTHCHECK and will do appropriate actions.

**Comment by bvis** [comment]
Ok, thanks. I thought that I always needed Consul as backend to have a persistent status that would allow me to start more tasks of the service.

On other hand I'd suggest to add another more option that would be to pass the TLS certificate data needed by a domain to work under HTTPS in a label as well.

Of course this could cause some security problems in case your cluster is compromised, because anybody could check the labels used on each service and could see the TLS data, but I think this could be very useful under many circumstances.

Another option, to not to need to build new images each time a new (or updated) certificate  is needed by any application of the swarm cluster, could be to integrate the certbot that uses LetsEncrypt to automatically manage the https certificates. But this could be done in another image extending this one, or another service connected to this, because it could make the image to do too many things. Probably this deserves another task or debate.

Thanks again for your work.

**Comment by pascalandy** [comment]
So +1 on this :)

**Comment by vfarcic** [comment]
There are a couple of requests in this issue. @bvis Would you mind breaking them down into separate tickets so that we can track them down and fix them one by one?

One of those, HEALTHCHECK, is done. It was very easy. The initial configuration is loaded before the Web server is started so all that was needed for the check is to send a request to the API. It's available in the release 1.90.

**Comment by bvis** [comment]
Of course. I'll clone this task asking you to support labels for automatic configuration of the services to expose.
